allow lalsimulation to be missing, helper function for optional libraries

ahnitz opened the pull request:
This aims to allow lalsimulation to be separated from some more basic PyCBC functionality. This PR should not reduce any existing functionality. 

Where possible, this just means that some functions will have fewer options if lalsimulation is not available (i.e. waveform generation, analytic psds). In other cases, if a function which requires lalsimulation is called (injections, some detector stuff, etc), I've tried to enable a useful error message. 

Noteably, I've added a helper function which may be useful for optional functionality in other cases. It allows one to import a module and not have to add extra try / expect statements everywhere for the common case where specific methods or functions need to call the module. 

one uses this as 

```
import pycbc.libutils
lalsim = pycbc.libutils.import_optional('lalsimulation')

def my_func(something, something):
      # use lalsim
      lalsim.something(something)
```

This function will not return an error, but if the lalsimulation is not importable it will raise an importerror with a message telling which function requires this module *only* when you actually use something provided by the module. This allows a cleaner syntax where you can still "import" the module at the top-level and use within functions as normal without adding extra error handling. If you need to use the module directly within another module (and not hidden by a function definition) then error handling is still of course required.

ahnitz commented:
I think this is now ready for broad review, I'll take a look at the CC parts tomorrow. I think most of them I can't do much about though. 

@cdcapano Just poking you to point out I changed how the import of pykerr works so that it works in the same was as lalsimulation now and the error message for being missing will be in the same form.

spxiwh commented on pycbc/detector.py:
```diff
@@ -176,7 +177,7 @@ def __init__(self, detector_name, reference_time=1126259462.0):
         self.name = str(detector_name)
 
         if detector_name in [pfx for pfx, name in get_available_detectors()]:
-            import lalsimulation as lalsim
```
Shouldn't this be moved to the top of the code?

spxiwh commented on pycbc/inject/inject.py:
```diff
@@ -46,13 +44,13 @@
 injection_func_map = {
-    np.dtype(float32): sim.SimAddInjectionREAL4TimeSeries,
-    np.dtype(float64): sim.SimAddInjectionREAL8TimeSeries
+    np.dtype(float32): lambda *args: sim.SimAddInjectionREAL4TimeSeries(*args),
```
I'm surprised that this wouldn't fail if lalsimulation is not present??

ahnitz replied:
It works because the lambda isn't actually evaluating the expression, whereas it would be if directly putting a function into a dictionary. It's a way to hide the use until it is used rather than when creating the dict.

spxiwh commented on pycbc/results/versioning.py:
```diff
@@ -82,7 +83,7 @@ def add_info_new_version(info_dct, curr_module, extra_str):
         lalsimulationinfo['Date'] = lalsimulation.SimulationVCSDate
-    except AttributeError:
+    except (AttributeError, ImportError):
         add_info_new_version(lalsimulationinfo, lalsimulation, 'Simulation')
```
Would this line work with lalsimulation not present?

spxiwh reviewed (commented):
I left a few minor comments, but I think this PR will do what it proposes to do.

I'm not sure if I think it's worth adding this complexity to let PyCBC work without lalsimulation being present (because I haven't really understood a clear motivation for doing this). There is probably more in PyCBC that depends on lalsimulation than certain other dependencies, and it's not clear why we would treat this as a special case, given that it is not difficult to install lalsimulation in most common use cases (indeed we've had many more problems with other dependencies than this one).

So I'm on the fence with this one right now.

ahnitz commented:
@spxiwh Thanks, I'll look at addressing your comments. The main reason in the short term is that we have a specific use case with some of our CW colleagues here. They would like to use parts of the pycbc library, but have alternates for parts of lalsuite. This does not include lalsimulation, however, and it's not easy to provide that. 

Longer term, I think lalsuite in general should be an optional dependency for ground-based specific tasks. That way we can treat lalsuite and equivalent LISA code on the same footing without requiring each other. The ideal is that we can provide the high level python interfaces for waveform / signal generation but can call into those for specific functionality.

spxiwh commented:
Okay, I think this has been open long enough for objections so I'll approve this now.
